src/mito2/src/read/series_scan.rs

```diff
@@ -627,3 +758,332 @@ fn new_partition_metrics(
     metrics_list.set(partition, metrics.clone());
     metrics
 }
+
+/// A divider to split flat record batches by time series.
+///
+/// It only ensures rows of the same series are returned in the same [FlatSeriesBatch].
+/// However, a [FlatSeriesBatch] may contain rows from multiple series.
+#[derive(Default)]
+struct FlatSeriesBatchDivider {
+    buffer: FlatSeriesBatch,
+}
+
+impl FlatSeriesBatchDivider {
+    /// Pushes a record batch into the divider.
+    ///
+    /// Returns a [FlatSeriesBatch] if the series in the buffer is exhausted.
+    fn push(&mut self, batch: RecordBatch) -> Option<FlatSeriesBatch> {
+        // If buffer is empty
+        if self.buffer.batches.is_empty() {
+            self.buffer.batches.push(batch);
+            return None;
+        }
+
+        // Gets the primary key column from the incoming batch.
+        let pk_column_idx = primary_key_column_index(batch.num_columns());
+        let batch_pk_column = batch.column(pk_column_idx);
+        let batch_pk_array = batch_pk_column
+            .as_any()
+            .downcast_ref::<PrimaryKeyArray>()
+            .unwrap();
+        let batch_pk_values = batch_pk_array
+            .values()
+            .as_any()
+            .downcast_ref::<BinaryArray>()
+            .unwrap();
+        // Gets the last primary key of the incoming batch.
+        let batch_last_pk =
+            primary_key_at(batch_pk_array, batch_pk_values, batch_pk_array.len() - 1);
+        // Gets the last primary key of the buffer.
+        // Safety: the buffer is not empty.
+        let buffer_last_batch = self.buffer.batches.last().unwrap();
+        let buffer_pk_column = buffer_last_batch.column(pk_column_idx);
+        let buffer_pk_array = buffer_pk_column
+            .as_any()
+            .downcast_ref::<PrimaryKeyArray>()
+            .unwrap();
+        let buffer_pk_values = buffer_pk_array
+            .values()
+            .as_any()
+            .downcast_ref::<BinaryArray>()
+            .unwrap();
+        let buffer_last_pk =
+            primary_key_at(buffer_pk_array, buffer_pk_values, buffer_pk_array.len() - 1);
+
+        // If last primary key of the batch is the same as last primary key of the buffer
+        if batch_last_pk == buffer_last_pk {
+            self.buffer.batches.push(batch);
+            return None;
+        }
+
+        // Finds the offset where the primary key changes.
+        let batch_pk_keys = batch_pk_array.keys();
+        let pk_indices = batch_pk_keys.values();
+        let mut change_offset = 0;
+        for (i, &key) in pk_indices.iter().enumerate() {
+            let batch_pk = batch_pk_values.value(key as usize);
+
+            if buffer_last_pk != batch_pk {
+                change_offset = i;
+                break;
+            }
```

Comment:
If all the primary keys are the same, should we expect the `change_offset` to be 0 as well?

Reply:
We have checked the last pk is not the same as the buffer pk in L814, so this is impossible.

Reply:
is this right?
```
let change_offset = pk_indices.iter()
      .enumerate()
      .find(|(_, &key)| {
          let batch_pk = batch_pk_values.value(key as usize);
          buffer_last_pk != batch_pk
      })
      .map(|(i, _)| i);
```

Reply:
got it